ext/standard/html.c

```diff
@@ -1559,3 +1560,52 @@ PHP_FUNCTION(get_html_translation_table)
 	}
 }
 /* }}} */
+
+PHP_FUNCTION(request_parse_body)
+{
+	zval *input_stream_zv = NULL;
+	zend_string *content_type = NULL;
+
+	ZEND_PARSE_PARAMETERS_START(0, 2)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_RESOURCE_OR_NULL(input_stream_zv)
+		Z_PARAM_STR_OR_NULL(content_type)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (!content_type && input_stream_zv) {
+		zend_throw_error(NULL, "$content_type must be set if $input_stream is passed");
+		RETURN_THROWS();
+	} else if (content_type && !input_stream_zv) {
+		zend_throw_error(NULL, "$content_type must not be set if $input_stream is not passed");
+		RETURN_THROWS();
+	}
+
+	php_stream *input_stream = NULL;
+	if (input_stream_zv) {
+		php_stream_from_zval(input_stream, input_stream_zv);
+	}
+	SG(rfc1867_input_stream) = input_stream;
```

Comment:
Can you pass ``input_stream`` as an argument of ``sapi_handle_post_ex()``? Or may be even combine ``sapi_read_post_data_ex()`` and ``sapi_handle_post_ex()`` together.

I may be wrong.

Reply:
We can avoid the global if we further adjust `post_handler` and `post_reader`. RFC1867 itself uses the input stream from `fill_buffer`, which will also require passing the variable through a few function boundaries. I'm not sure whether that's cleaner.

> Or may be even combine sapi_read_post_data_ex() and sapi_handle_post_ex() together.

`post_read` is used for preemptively reading the content of any POST request except for `multipart/form-data` into `php://input`. `post_handler` is used for parsing `application/x-www-form-urlencoded` or multipart. urlencoded reads from php://input, and multipart (before this PR) reads directly from SAPI. The distinction of read and handle is rather useless, as `urlencoded` could simply call into `post_read` manually.

That said, I think it may be better to make improvements to the API after merging these changes in case the RFC is accepted. I will see if I can minimize the changes here.